### PR TITLE
XLA train step fixes

### PR DESCRIPTION
**Pull request by Rocketknight1**
This PR makes a bunch of changes to the TF codebase to improve XLA support, in preparation for our upcoming big TF release. The goal is to allow users to use `jit_compile` on the vast majority of our models, which should yield large performance improvements for TF. In particular:

- Rewrites to the `train_step` and `test_step` so that any mutable Python input dicts are not modified in the step. This was a bad idea anyway, but it causes particular problems with XLA, which is very functional and hates side effects, like JAX.
- Rewrites to the common `hf_compute_loss` functions to ensure that static shapes are maintained throughout, so that XLA compilation is possible.
- Add a test to ensure that we can still fit models when XLA compilation is used. XLA compilation is quite expensive, which makes this test quite slow, so it's restricted to `core` models for now and tagged as `@slow`.

Left to do:
- [x] Fix XLA-incompatible model-specific  `hf_compute_loss` functions. On a quick search it looked like there were 4-5 of these, so it shouldn't take too long. Any use of `tf.boolean_mask` is a surefire sign that XLA compilation will break, because output shapes become data-dependent.
- [x] See if there's a way to test non-core models for XLA fit support without crippling performance. (No, but we're using the XLA losses in non-XLA tests by default, so that partially tests it for all models)

### CODE REVIEW

**Comment by LysandreJik** [comment]
I'd be interested in having @ydshieh's review as well

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -196,10 +196,13 @@ def hf_compute_loss(self, labels, logits):
...
         )
         # make sure only labels that are not equal to -100 affect the loss
-        active_loss = tf.not_equal(tf.reshape(labels, (-1,)), -100)
-        reduced_logits = tf.boolean_mask(tf.reshape(logits, (-1, shape_list(logits)[2])), active_loss)
-        labels = tf.boolean_mask(tf.reshape(labels, (-1,)), active_loss)
-        return loss_fn(labels, reduced_logits)
+        unmasked_loss = loss_fn(labels, logits)
+        loss_mask = tf.cast(labels != -100, dtype=unmasked_loss.dtype)
+        loss_denominator = tf.reduce_sum(loss_mask, axis=1)
+        # Masked positions will have a loss of NaN because -100 is not a valid label
```
Comment (ydshieh) [inline src/transformers/modeling_tf_utils.py:202]: Is this documented somewhere in TF doc? I don't remember this behavior.

**Review by ydshieh — COMMENTED** [review]
@Rocketknight1 Great for more XLA compatibility!

Haven't checked the changes in test files yet, but left a few comments/questions in `modeling_tf_utils.py`

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -196,10 +196,13 @@ def hf_compute_loss(self, labels, logits):
...
+        masked_loss = tf.math.multiply_no_nan(unmasked_loss, loss_mask)
+        reduced_masked_loss = tf.reduce_sum(masked_loss, axis=1) / loss_denominator
+        return reduced_masked_loss
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/modeling_tf_utils.py:205]: I was checking the docs now, and PT only outputs one number for the batch -- so it makes sense to include the sum here 👍 

(we would have to update the TF docstring)

File: src/transformers/modeling_tf_utils.py [hunk 3]
```diff
@@ -251,7 +255,7 @@ class TFSequenceClassificationLoss:
     """
 
     def hf_compute_loss(self, labels, logits):
-        if len(shape_list(logits)) == 1 or shape_list(logits)[1] == 1:
+        if logits.shape.rank == 1 or logits.shape[1] == 1:
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/modeling_tf_utils.py:258]: I also don't remember the exact issue, other than it often causes XLA compilation to fail :)

File: src/transformers/modeling_tf_utils.py [hunk 4]
```diff
@@ -300,11 +304,13 @@ def hf_compute_loss(self, labels, logits):
...
-        next_sentence_reduced_logits = tf.boolean_mask(tf.reshape(logits, (-1, 2)), next_sentence_active_loss)
-        next_sentence_label = tf.boolean_mask(tf.reshape(labels, (-1,)), next_sentence_active_loss)
 
-        return loss_fn(next_sentence_label, next_sentence_reduced_logits)
+        unmasked_ns_loss = loss_fn(y_true=labels, y_pred=logits)
+        ns_loss_mask = tf.cast(labels != -100, dtype=unmasked_ns_loss.dtype)
+        # Just zero out samples where label is -100, no reduction
+        masked_ns_loss = tf.math.multiply_no_nan(unmasked_ns_loss, ns_loss_mask)
+
+        return masked_ns_loss
```
Comment (gante) [inline src/transformers/modeling_tf_utils.py:313]: Should this one be reduced as well?

**Review by gante — COMMENTED** [review]
XLA <3

File: src/transformers/modeling_tf_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/modeling_tf_utils.py:205]: (BTW, the PT output the average instead of the sum, as I see in `GPT2LMHeadModel` - `CrossEntropyLoss` have `reduction='mean'` by default).

File: src/transformers/modeling_tf_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/modeling_tf_utils.py:258]: I didn't see `tf.shape` used in the previous version.

Regarding `.shape` v.s `shape_list`, I am ok as long as things work. I  am just feel confused that most of the time (in other places) I see `.shape` fails while `shape_list` works (with graph mode / symbolic tensors etc.).

File: src/transformers/modeling_tf_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:202]: I don't know if it's documented, but any cross-entropy loss will yield `NaN` or `-inf` when the model's probability for the true label is 0. Since labels < 0 do not correspond to a valid category, the loss in those cases will always be one of those values.

File: src/transformers/modeling_tf_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:258]: Ah, sorry, I should explain! `shape_list` uses a combination of `.shape` and `tf.shape()` to build the list. I think using `tf.shape()` here confuses XLA, because it looks like a conditional that depends on the specific data you input, and those are forbidden.

I'm not 100% sure of the exact rules it uses, but all I can tell you is that it failed before and it works like this!

File: src/transformers/modeling_tf_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:205]: It will be impossible to keep the old behaviour with XLA, because the number of 'active' tokens will change in each batch.

We could return a scalar number, but in Keras it's nice to return a vector of per-sample losses, because this means the user can use the `sample_weight` argument to `fit()` if they want to. I think that's fairly uncommon though, so if we want to stick with a scalar, that's fine!

File: src/transformers/modeling_tf_utils.py (hunk 4 above)
Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:313]: `ns_loss` is calculated per sample rather than per position, so `masked_ns_loss` is a vector of shape `(num_samples,)`. We could reduce that to a scalar if we want, though!

**Review by sgugger — COMMENTED** [review]
If I understand correctly, this changes the loss returned by TensorFlow models from a matrix to a vector, which is obviously breaking. While I don't know how many TensorFlow users rely on the current structure of the loss, we at least need to have a flag (probably `use_xla=False`) to enable the previous behavior for users who relied on it.

Could you confirm first that my understanding is correct?

**Comment by ydshieh** [comment]
> If I understand correctly, this changes the loss returned by TensorFlow models from a matrix to a vector, which is obviously breaking. While I don't know how many TensorFlow users rely on the current structure of the loss, we at least need to have a flag (probably `use_xla=False`) to enable the previous behavior for users who relied on it.
> 
> Could you confirm first that my understanding is correct?

I believe both prev. and current version returns a vector. The difference is on the size: 

- prev: number of active tokens (non-padding tokens)
- now: batch size

File: src/transformers/configuration_utils.py [hunk 5]
```diff
@@ -236,6 +236,9 @@ class PretrainedConfig(PushToHubMixin):
 
         use_bfloat16 (`bool`, *optional*, defaults to `False`):
             Whether or not the model should use BFloat16 scalars (only used by some TensorFlow models).
+        tf_legacy_loss (`bool`, *optional*, defaults to `False`):
+            Whether or not the model should use legacy TensorFlow losses. Legacy losses have variable output
+            shapes and may not be XLA-compatible.
```
Comment (sgugger) [inline src/transformers/configuration_utils.py:241]: Maybe add that this parameter is here for backward compatibility but will be removed in v5?

File: src/transformers/modeling_tf_utils.py [hunk 6]
```diff
@@ -195,11 +195,21 @@ def hf_compute_loss(self, labels, logits):
         loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(
             from_logits=True, reduction=tf.keras.losses.Reduction.NONE
         )
+        if self.config.get("tf_legacy_loss", False):
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:198]: Configs are not dict, this will fail.
```suggestion
        if self.config.tf_legacy_loss:
```
(Same below)

Comment (Rocketknight1) [reply src/transformers/modeling_tf_utils.py:198]: Ah, sorry, I meant to use `getattr`, but I guess the key will always be present anyway?

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

File: src/transformers/configuration_utils.py (hunk 5 above)
Comment (Rocketknight1) [reply src/transformers/configuration_utils.py]: Done!

**Comment by Rocketknight1** [comment]
@ydshieh I was completely wrong earlier - `SparseCategoricalCrossentropy` only returns `nan` for invalid labels when running on GPU! On CPU, inputs are validated and TensorFlow throws an error.

I'll rewrite my loss functions to not depend on that behaviour, and change the loss computation tests to mask some positions to ensure that gets tested, so I don't miss anything like this in future.

**Comment by ydshieh** [comment]
@Rocketknight1 

https://github.com/huggingface/transformers/blob/f17136c80dfec2a78890d012105634079531dcd9/src/transformers/modeling_tf_utils.py#L213

As in an earlier comment, I think this loss value is incorrect. Image we have 2 sequences of length 100.
- 1st sentence: 1 active token + 99 pad tokens (somehow non-sense 😄 )
- 2nd sentence: 20 active token + 80 pad tokens

In this latest version, the unique token in sentence 1 get an weight (when computing the loss) 20 times larger than each token in the 2nd sentence. (As you first average the loss along sequence dimension).

Furthermore, this doesn't correspond to PyTorch's computation, which leads to test failures (I didn't check in detail if this is the cause, but I believe it is).

**Q: Is there any reason we don't want to sum each token's loss value?**

cc @gante  @patrickvonplaten

**Comment by Rocketknight1** [comment]
Hi @ydshieh I'm sorry, I think you're right there! Let me investigate and see if I can make a PR to weight tokens properly, which should hopefully resolve the issue.

**Review by patrickvonplaten — COMMENTED** [review]
Big +1 on @ydshieh's comment here. I think the average doesn't correctly consider different padding lengths

**Comment by Rocketknight1** [comment]
@patrickvonplaten Agreed! I fixed that in https://github.com/huggingface/transformers/pull/18013

File: src/transformers/configuration_utils.py [hunk 7]
```diff
@@ -236,6 +236,10 @@ class PretrainedConfig(PushToHubMixin):
 
         use_bfloat16 (`bool`, *optional*, defaults to `False`):
             Whether or not the model should use BFloat16 scalars (only used by some TensorFlow models).
+        tf_legacy_loss (`bool`, *optional*, defaults to `False`):
```
Comment (patrickvonplaten) [inline src/transformers/configuration_utils.py:239]: Are we sure we want to set the default to `False`? This is breaking no? Also it's a somewhat hard-to-discover silent error in this case no?

Comment (Rocketknight1) [reply src/transformers/configuration_utils.py:239]: It's only very slightly breaking - anyone using Keras or a custom model will not notice any change. The existing losses return very strange shapes like vectors of shape `(num_unmasked_tokens,)` that vary in length each iteration, with no mapping from there back to the original tokens. I doubt anyone is using them directly.
